src/cpuinfo.cpp

```diff
@@ -25,452 +27,721 @@
-#define RISCV_HWPROBE_EXT_ZALRSC        (1ULL << 57)
+#define RISCV_HWPROBE_EXT_ZALRSC (1ULL << 57)
 #endif
 
-#ifndef COMPAT_HWCAP_ISA_I
-#define COMPAT_HWCAP_ISA_I  (1U << ('I' - 'A'))
-#endif
+namespace {
 
-#ifndef COMPAT_HWCAP_ISA_M
-#define COMPAT_HWCAP_ISA_M  (1U << ('M' - 'A'))
-#endif
+#if defined(__linux__) && defined(__riscv)
+bool UseSigillHandler(biscuit::RISCVExtension extension) {
+    using namespace biscuit;
+    switch (extension) {
+    case RISCVExtension::I:
+    case RISCVExtension::M:
+    case RISCVExtension::A:
+    case RISCVExtension::F:
+    case RISCVExtension::D:
+    case RISCVExtension::C:
+    case RISCVExtension::V:
+    case RISCVExtension::Zba:
+    case RISCVExtension::Zbb:
+    case RISCVExtension::Zbs:
+    case RISCVExtension::Zicboz:
+    case RISCVExtension::Zbc:
+    case RISCVExtension::Zbkb:
+    case RISCVExtension::Zbkc:
+    case RISCVExtension::Zbkx:
+    case RISCVExtension::Zknd:
+    case RISCVExtension::Zkne:
+    case RISCVExtension::Zknh:
+    case RISCVExtension::Zksed:
+    case RISCVExtension::Zksh:
+    case RISCVExtension::Zvbb:
+    case RISCVExtension::Zvbc:
+    case RISCVExtension::Zvkb:
+    case RISCVExtension::Zvkg:
+    case RISCVExtension::Zvkned:
+    case RISCVExtension::Zvknha:
+    case RISCVExtension::Zvknhb:
+    case RISCVExtension::Zvksed:
+    case RISCVExtension::Zvksh:
+    case RISCVExtension::Zvkt:
+    case RISCVExtension::Zfh:
+    case RISCVExtension::Zfhmin:
+    case RISCVExtension::Zvfh:
+    case RISCVExtension::Zvfhmin:
+    case RISCVExtension::Zfa:
+    case RISCVExtension::Zacas:
+    case RISCVExtension::Zicond:
+    case RISCVExtension::Zicntr:
+    case RISCVExtension::Zihpm:
+    case RISCVExtension::Zfbfmin:
+    case RISCVExtension::Zicbom: {
+        return true;
+    }
+    default: {
+        return false;
+    }
+    }
+}
 
-#ifndef COMPAT_HWCAP_ISA_A
-#define COMPAT_HWCAP_ISA_A  (1U << ('A' - 'A'))
-#endif
+void SigillHandler(int sig, struct siginfo_t* info, void* ctx) {
+    mcontext_t* mctx = &((ucontext_t*)ctx)->uc_mcontext;
 
-#ifndef COMPAT_HWCAP_ISA_F
-#define COMPAT_HWCAP_ISA_F  (1U << ('F' - 'A'))
-#endif
+    // Since we hit SIGILL, set return value to false
+    mctx->__gregs[REG_A0] = 0;
 
-#ifndef COMPAT_HWCAP_ISA_D
-#define COMPAT_HWCAP_ISA_D  (1U << ('D' - 'A'))
-#endif
+    // Instead of incrementing the PC, just set it to the return value immediately
+    // after we return from the signal handler. This way we don't need to check if
+    // it's a compressed instruction or not to find out how much to increment PC by.
+    // Additionally, some extensions may require checking multiple instructions.
+    mctx->__gregs[REG_PC] = mctx->__gregs[REG_RA];
+}
 
-#ifndef COMPAT_HWCAP_ISA_C
-#define COMPAT_HWCAP_ISA_C  (1U << ('C' - 'A'))
-#endif
+void SetEGW(biscuit::Assembler& as, biscuit::SEW eew, uint32_t egs) {
+    using namespace biscuit;
+    uint32_t vlen = CPUInfo().GetVlenb() * 8;
+    uint32_t egw;
```

Comment:
Should be initialized, especially if we allow redefinition of `BISCUIT_ASSERT` in the future.